### ISSUE TITLE
Unstable test - StackChoiceViewTest.test_sync_roi_plot_for_old_stack_with_new_stack

### ISSUE THREAD

**Issue by dtasev** [mantidproject/mantidimaging#772]
Full stack trace

```
=================================================================================== FAILURES ====================================================================================
______________________________________________________ StackChoiceViewTest.test_sync_roi_plot_for_old_stack_with_new_stack ______________________________________________________

self = <mantidimaging.gui.windows.stack_choice.tests.test_view.StackChoiceViewTest testMethod=test_sync_roi_plot_for_old_stack_with_new_stack>

    def setUp(self):
        self.original_stack = th.generate_images()
        self.new_stack = th.generate_images()
        self.p = mock.MagicMock()
>       self.v = StackChoiceView(self.original_stack, self.new_stack, self.p, None)

mantidimaging/gui/windows/stack_choice/tests/test_view.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mantidimaging/gui/windows/stack_choice/view.py:47: in __init__
    self.new_stack = MIImageView(detailsSpanAllCols=True)
mantidimaging/gui/widgets/mi_image_view/view.py:41: in __init__
    super().__init__(parent, name, view, imageItem, levelMode, *args)
../../apps/miniconda3/envs/mitest/lib/python3.7/site-packages/pyqtgraph/imageview/ImageView.py:211: in __init__
    self.view.register(self.name)
../../apps/miniconda3/envs/mitest/lib/python3.7/site-packages/pyqtgraph/graphicsItems/ViewBox/ViewBox.py:263: in register
    ViewBox.updateAllViewLists()
../../apps/miniconda3/envs/mitest/lib/python3.7/site-packages/pyqtgraph/graphicsItems/ViewBox/ViewBox.py:1654: in updateAllViewLists
    for v in ViewBox.AllViews:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <WeakKeyDictionary at 0x7fa365fca710>

    def keys(self):
        with _IterationGuard(self):
>           for wr in self.data:
E           RuntimeError: dictionary changed size during iteration

../../apps/miniconda3/envs/mitest/lib/python3.7/weakref.py:451: RuntimeError
```

**Comment by DolicaAkelloEgwel** [comment]
Any idea what conditions cause this test to fail?

**Comment by dtasev** [comment]
Possibly because they all run in parallel, so sometimes we get a race condition. Perhaps we ought to be using `pytest-xdist` for GUI tests, and `pytest-parallel` for CORE tests, although I've no guarantee that will fix anything, the difference is (from the pytest pip https://pypi.org/project/pytest-parallel/):

```
pytest-xdist is great to run tests that:

- aren't threadsafe
- perform poorly when multithreaded
- need state isolation

pytest-parallel is better for some use cases (like Selenium tests) that:

- can be threadsafe
- can use non-blocking IO for http requests to make it performant
- manage little or no state in the Python environment

Put simply, `pytest-xdist` does parallelism while `pytest-parallel` does parallelism and concurrency.
```

Possibly the issue is caused by the concurrency, and some process changing the view while another test is running on it

**Comment by samtygier-stfc** [comment]
I've run the tests in a loop for a few hours without triggering this on master.

It possible that either the switch to python 3.8 or switch from pytest-parallel -> pytest-xdist has fixed this. I'll close, but we can reopen if it shows up again.
